### DISCUSSION TITLE
OpenAPI generation fails with dates with Annotated constraint

### DISCUSSION THREAD

**Discussion by jereflit** [fastapi/fastapi#11687] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from datetime import date
from typing import Annotated

import annotated_types
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class DataInput(BaseModel):
    birthdate: Annotated[
        date,
        annotated_types.Gt(date(1900, 1, 1)),
        annotated_types.Lt(date(2000, 1, 1)),
    ]


@app.post("/")
def post_root(
    data_input: DataInput,
):
    return {"birthdate": data_input.birthdate}
```
EDIT: the `birthdate` field must be a `date` field of a pydantic model, as is it used in a POST endpoint

### Description

A pydantic ValidationError arises when FastApi calls the openapi utils that generates an instance of `OpenAPI` https://github.com/tiangolo/fastapi/blob/a9819dfd8da39a754837cc134df4aca6c0a9a3f6/fastapi/openapi/utils.py#L530 

```
  ...
  File "/home/jereflit/bug/.venv/lib/python3.11/site-packages/fastapi/applications.py", line 1009, in openapi
    return JSONResponse(self.openapi())
                        ^^^^^^^^^^^^^^
  File "/home/jereflit/bug/.venv/lib/python3.11/site-packages/fastapi/applications.py", line 981, in openapi
    self.openapi_schema = get_openapi(
                          ^^^^^^^^^^^^
  File "/home/jereflit/bug/.venv/lib/python3.11/site-packages/fastapi/openapi/utils.py", line 530, in get_openapi
    return jsonable_encoder(OpenAPI(**output), by_alias=True, exclude_none=True)  # type: ignore
                            ^^^^^^^^^^^^^^^^^
  File "/home/jereflit/bug/.venv/lib/python3.11/site-packages/pydantic/main.py", line 176, in __init__
    self.__pydantic_validator__.validate_python(data, self_instance=self)

pydantic_core._pydantic_core.ValidationError: 4 validation errors for OpenAPI
components.schemas.PricerInput.Schema.properties.driver_birth_date.Schema.exclusiveMaximum
  Input should be a valid number [type=float_type, input_value=datetime.date(2006, 6, 5), input_type=date]
    For further information visit https://errors.pydantic.dev/2.7/v/float_type
components.schemas.PricerInput.Schema.properties.driver_birth_date.Schema.exclusiveMinimum
  Input should be a valid number [type=float_type, input_value=datetime.date(2005, 6, 6), input_type=date]
    For further information visit https://errors.pydantic.dev/2.7/v/float_type
components.schemas.PricerInput.Schema.properties.driver_birth_date.bool
  Input should be a valid boolean [type=bool_type, input_value={'description': 'Date of ...Date', 'type': 'string'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.7/v/bool_type
components.schemas.PricerInput.Reference.$ref
  Field required [type=missing, input_value={'properties': {'bonus_ma...nput', 'type': 'object'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.7/v/missing
```

Note that the concerned fields are indeed defined as optional floats : https://github.com/tiangolo/fastapi/blob/a9819dfd8da39a754837cc134df4aca6c0a9a3f6/fastapi/openapi/models.py#L164-L167 because OpenAPI is not supposed to support values restriction for dates, which is confirmed here : https://github.com/OAI/OpenAPI-Specification/discussions/2827

However, FastAPI raises a `ValidationError` and makes `/openapi.json` fails, and therefore Swagger `/docs` and Redoc `/redoc` fail too, whereas the annotated type is correctly defined.

In the OpenAPi discussion linked above, someone suggests to use custom keywords `minimumDate` and `maximumDate`.
👉 Is it something that I can explicitly do on my side by writing my own `get_openapi` function ?

(A suggested workaround would be to handle the validation myself by adding a pydantic's `@field_validator` to my `DataInput` class. It is fine, except that neither OpenAPI json nor the type checker will reflect those constraints on input data.)

### Operating System

Linux

### FastAPI Version

0.111.0

### Pydantic Version

2.7.3

### Python Version

Python 3.11.5

**Comment by JavierSanchezCastro** [comment]
With `datetime` you can convert it to timestamp and set the gt and lt in the Query param

```python3
from datetime import date, datetime
from typing import Annotated

from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/")
def read_root(
    data_input: Annotated[date, Query(gt=datetime(1900, 1, 1).timestamp(), lt=datetime(2000, 1, 1).timestamp())]
):
    return {"birthdate": data_input}
```

**Reply by jereflit** [reply]
Thanks for your help !

It does work like in your example (even with a regular `date`):

```python
from datetime import date
from typing import Annotated

import annotated_types
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def read_root(
    birthdate: Annotated[date, annotated_types.Gt(date(1900, 1, 1)), annotated_types.Lt(date(2000, 1, 1))],
):
    return {"birthdate": birthdate}

```
but this is because `birthdate` is defined here as a Query param, which seems to be supported by the OpenAPI generator :
```json
...
"parameters": [
    {
        "name": "birthdate",
        "in": "query",
        "required": true,
        "schema": {
            "type": "string",
            "format": "date",
            "exclusiveMaximum": "2000-01-01",
            "exclusiveMinimum": "1900-01-01",
            "title": "Birthdate"
        }
    }
],
...
```

However, in my case, `birthdate` is a field of a Body param defined in a pydantic model, and needs to remain that way.

**Reply by JavierSanchezCastro** [reply, marked as answer]
Well if it's a Body param you need to use POST instead of GET. Taking that in mind, you can rewrite the code as

```python3
from datetime import date, datetime
from typing import Annotated
from pydantic import BaseModel, Field
from fastapi import FastAPI

app = FastAPI()

class DataInput(BaseModel):
    birthdate: Annotated[date, Field(gt=datetime(1900, 1, 1).timestamp(), lt=datetime(2000, 1, 1).timestamp())]

@app.post("/")
def read_root(
    data_input: DataInput
):
    return {"birthdate": data_input.birthdate}
```

**Reply by jereflit** [reply]
My bad, it should have been a POST from the beginning indeed.

However, what you suggest does not work, because `datetime` is not compatible with `date` : 

```
pydantic_core._pydantic_core.SchemaError: Invalid Schema:
model.schema.model-fields.fields.birthdate.schema.date.lt
  Datetimes provided to dates should have zero time - e.g. be exact dates [type=date_from_datetime_inexact, input_value=946681200.0, input_type=float]
    For further information visit https://errors.pydantic.dev/2.7/v/date_from_datetime_inexact
model.schema.model-fields.fields.birthdate.schema.date.gt
  Datetimes provided to dates should have zero time - e.g. be exact dates [type=date_from_datetime_inexact, input_value=-2208989361.0, input_type=float]
    For further information visit https://errors.pydantic.dev/2.7/v/date_from_datetime_inexact
```
unless as I accept birthdate to be given as datetimes... but this is not my use case.

**Reply by JavierSanchezCastro** [reply]
I've tried with your versions and it's working well

`pydantic==2.7.3`
`fastapi==0.111.0`

The type for birthdate is `date`, but just for the verification I used `datetime`

![imagen](https://github.com/tiangolo/fastapi/assets/72013291/99a0f581-74a7-4de9-bb46-45601d4ded6f)
![imagen](https://github.com/tiangolo/fastapi/assets/72013291/60c6f4e4-c36d-4c84-9584-aa69c399141f)

**Reply by jereflit** [reply]
I don't understand. I cannot even launch the api with your code : 
```bash
 ~/bug   uvicorn main:app --reload
```
``` bash

INFO:     Will watch for changes in these directories: ['/home/jereflit/bug']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [175641] using WatchFiles
Process SpawnProcess-1:
Traceback (most recent call last):
  File "/home/jereflit/.pyenv/versions/3.11.5/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  
  [.....skipped.....]

  File "/home/jereflit/bug/main.py", line 8, in <module>
    class DataInput(BaseModel):
  File "/home/jereflit/bug/.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 202, in __new__
    complete_model_class(
  File "/home/jereflit/bug/.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 549, in complete_model_class
    schema = gen_schema.clean_schema(schema)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jereflit/bug/.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py", line 442, in clean_schema
    schema = validate_core_schema(schema)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jereflit/bug/.venv/lib/python3.11/site-packages/pydantic/_internal/_core_utils.py", line 568, in validate_core_schema
    return _validate_core_schema(schema)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.SchemaError: Invalid Schema:
model.schema.model-fields.fields.birthdate.schema.date.lt
  Datetimes provided to dates should have zero time - e.g. be exact dates [type=date_from_datetime_inexact, input_value=946681200.0, input_type=float]
    For further information visit https://errors.pydantic.dev/2.7/v/date_from_datetime_inexact
model.schema.model-fields.fields.birthdate.schema.date.gt
  Datetimes provided to dates should have zero time - e.g. be exact dates [type=date_from_datetime_inexact, input_value=-2208989361.0, input_type=float]
    For further information visit https://errors.pydantic.dev/2.7/v/date_from_datetime_inexact
```

and yet I just copied/pasted it

**Reply by JavierSanchezCastro** [reply]
I can't reproduce your error with the versions you provide

**Reply by jereflit** [reply]
Here is my requirements.txt:

```requirements.txt
fastapi==0.111.0
uvicorn
```
et the result of `pip freeze`:
```
annotated-types==0.7.0
anyio==4.4.0
certifi==2024.6.2
click==8.1.7
dnspython==2.6.1
email_validator==2.1.1
fastapi==0.111.0
fastapi-cli==0.0.4
h11==0.14.0
httpcore==1.0.5
httptools==0.6.1
httpx==0.27.0
idna==3.7
Jinja2==3.1.4
markdown-it-py==3.0.0
MarkupSafe==2.1.5
mdurl==0.1.2
orjson==3.10.3
pydantic==2.7.3
pydantic_core==2.18.4
Pygments==2.18.0
python-dotenv==1.0.1
python-multipart==0.0.9
PyYAML==6.0.1
rich==13.7.1
shellingham==1.5.4
sniffio==1.3.1
starlette==0.37.2
typer==0.12.3
typing_extensions==4.12.1
ujson==5.10.0
uvicorn==0.30.1
uvloop==0.19.0
watchfiles==0.22.0
websockets==12.0
```

would you mind checking that you have the same please :pray: ? So I can change my packages accordingly. Thanks

**Reply by JavierSanchezCastro** [reply]
I created a docker image with your pip freeze result and the same python version and it still works (this example https://github.com/tiangolo/fastapi/discussions/11687#discussioncomment-9680826), try reinstalling pydantic and pydantic_core

**Reply by jereflit** [reply]
It would help if you could provide the dockerfile please, as I still cannot reproduce what is working on your side (even after reinstalling a virtualenv and re-doing `pip install -r requirements.txt`).

**Reply by JavierSanchezCastro** [reply]
Dockerfile

```dockerfile
FROM python:3.11.5-slim

WORKDIR /app

COPY requirements.txt .

RUN pip install --no-cache-dir -r requirements.txt

COPY . .

EXPOSE 8000

CMD ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "8000"]
```

main.py
```python3
from datetime import date, datetime
from typing import Annotated
from pydantic import BaseModel, Field
from fastapi import FastAPI

app = FastAPI()

class DataInput(BaseModel):
    birthdate: Annotated[date, Field(gt=datetime(1900, 1, 1).timestamp(), lt=datetime(2000, 1, 1).timestamp())]

@app.post("/")
def read_root(
    data_input: DataInput
):
    return {"birthdate": data_input.birthdate}
```

requirements.txt

```
annotated-types==0.7.0
anyio==4.4.0
certifi==2024.6.2
click==8.1.7
dnspython==2.6.1
email_validator==2.1.1
fastapi==0.111.0
fastapi-cli==0.0.4
h11==0.14.0
httpcore==1.0.5
httptools==0.6.1
httpx==0.27.0
idna==3.7
Jinja2==3.1.4
markdown-it-py==3.0.0
MarkupSafe==2.1.5
mdurl==0.1.2
orjson==3.10.3
pydantic==2.7.3
pydantic_core==2.18.4
Pygments==2.18.0
python-dotenv==1.0.1
python-multipart==0.0.9
PyYAML==6.0.1
rich==13.7.1
shellingham==1.5.4
sniffio==1.3.1
starlette==0.37.2
typer==0.12.3
typing_extensions==4.12.1
uvicorn==0.30.1
ujson==5.10.0
uvloop==0.19.0
watchfiles==0.22.0
websockets==12.0
```

## Commands

``sudo docker build -t fastapi_111687 .``
``sudo docker run -p 8000:8000 fastapi_111687``

Docker version 26.1.4, build 5650f9b

**Reply by jereflit** [reply]
Thanks a lot for your help and patience. Just like you, I can properly run your suggested solution in the docker container.

Therefore my local configuration is somehow different from the one in the docker container (and yet, in my virtual env, I have the same python version, and the same output of `pip freeze`). I have no clue about that.

I will accept your suggestion as the answer, even if I cannot use it locally :/

**Reply by JavierSanchezCastro** [reply]
Yeah, it seems a bit weird

**Comment by alessio-locatelli** [comment]
@JavierSanchezCastro Thanks. That was an awesome and useful support.

I did not like the idea to change a valid Pydantic code just to make FastAPI/OpenAPI to be happy, so I solved the problem by overwriting a schema:

```py
min_date = date(1900, 1, 1)

class DataInput(BaseModel):
    birthdate: Annotated[
        date,
        annotated_types.Ge(min_date),
    ] = Field(
            # I write `True` because OpenAPI asks for an `int` instance, but
            # you can put a min timestamp here. Actually, any value here makes
            # no sense because we accept a string like "YYYY-MM-DD".
            # https://github.com/OAI/OpenAPI-Specification/discussions/2827
            json_schema_extra={"minimum": True}
        )   


@router.post("/date")
def post_date(
    data_input: DataInput,
):
    return {"birthdate": data_input.birthdate}
```

The advantage of this solution is that you do not need to change a valid model.

The disadvantage is that you see in your API docs that a field accepts a string that is `>1` which is meaningless, but also OpenAPI already shows a valid example for `Request body` out-of-the-box:
```
{
  "birthdate": "2024-07-24"
}
```
so I see no significant problems with the API documentation (the API users will understand what is required to pass).
